File: src/reliable_transmission.erl
-module(reliable_transmission).

-export([send_data/2, receive_data/1]).

-define(RETRY_LIMIT, 5).
-define(ACK_TIMEOUT, 1000).

% Send data to a peer with retry mechanism
send_data(Peer, Data) ->
    send_data(Peer, Data, 0).

% Handle retry limit reached for sending data
send_data(_Peer, _Data, ?RETRY_LIMIT) ->
    error_logger:error_msg("Retry limit reached for sending data to peer: ~p~n", [_Peer]),
    {error, retry_limit_reached};
send_data(Peer, Data, RetryCount) ->
    % Send data to peer
    Peer ! {data, Data},
    % Wait for acknowledgment
    receive
        {ack, Peer} ->
            {ok, data_sent};
        after ?ACK_TIMEOUT ->
            % Retry sending data
            send_data(Peer, Data, RetryCount + 1)
    end.

% Receive data and process it using the provided data handler
receive_data(DataHandler) ->
    receive
        {data, Data} ->
            % Process received data
            DataHandler(Data),
            % Send acknowledgment
            self() ! {ack, self()},
            receive_data(DataHandler)
    end.

File: src/scala/main/scala/reliable_transmission.scala
import scala.concurrent.{Future, ExecutionContext}
import scala.concurrent.duration._
import scala.util.{Success, Failure}

object ReliableTransmission {
  implicit val ec: ExecutionContext = ExecutionContext.global

  private val retryLimit = 5
  private val ackTimeout = 1000

  def sendData(peer: String, data: String): Future[Unit] = {
    def attemptSend(retryCount: Int): Future[Unit] = {
      if (retryCount >= retryLimit) {
        Future.failed(new Exception("Retry limit reached"))
      } else {
        // Simulate sending data to peer
        println(s"Sending data to $peer: $data")
        // Simulate waiting for acknowledgment
        Future {
          Thread.sleep(ackTimeout)
          if (scala.util.Random.nextBoolean()) {
            println(s"Acknowledgment received from $peer")
            Success(())
          } else {
            println(s"No acknowledgment from $peer, retrying...")
            Failure(new Exception("No acknowledgment"))
          }
        }.flatMap {
          case Success(_) => Future.successful(())
          case Failure(_) => attemptSend(retryCount + 1)
        }
      }
    }

    attemptSend(0)
  }

  def receiveData(dataHandler: String => Unit): Future[Unit] = Future {
    // Simulate receiving data
    val data = "Received data"
    println(data)
    dataHandler(data)
    // Simulate sending acknowledgment
    println("Sending acknowledgment")
  }

  def handleTransmissionError(peer: String, reason: String): Unit = {
    println(s"Transmission error with peer $peer: $reason")
  }
}
